### DISCUSSION TITLE
Telemetry instrumentation vs Health Checking

### DISCUSSION THREAD

**Discussion by julealgon** [open-telemetry/opentelemetry-dotnet#4134] [General]
I've been using these OTEL instrumentation libraries a lot recently in conjunction with the OTEL Collector and Datadog as the monitoring tool.

Over the past few days, I started to take a look into Health Checking in AspNetCore, through this page:
- [Health checks in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0)

After reading a bit on this, and then taking a look at [the various health check-related packages](https://www.nuget.org/packages?q=healthchecks&sortBy=relevance) I'm a bit conflicted: there seems to be quite a bit of conceptual overlap between telemetry/metrics instrumentation, and these health checking procedures.

This is not so much a question but a discussion. How does the team/community view this potential overlap? Which one will "win", or will we end up using both mechanisms at the same time in the future?

The way I'm seeing these health checking packages is that they look a lot like OpenTelemetry instrumentation packages to me. Usually, its one package per "technology", and it produces some sort of "metric" or a communication result that serves to signal that the service is ok.

The healthchecking system even includes the "publisher" abstraction: a hosted service that pushes health data from time to time to some receiver (again, _very_ similar to me to what an OTEL exporter would do).

I also imagine one would want to see "application health" in the same tool that is used for monitoring logs/traces/metrics, at least in most situations, so there is a conceptual overlap in terms of exporting the data as well.

Should we even use both systems at the same time? Is there some guideline or recommendation in regards to the interaction between OTEL telemetry instrumentation/exporting vs this HealthCheck mechanism?

As it stands, the OTEL instrumentation can't be used for the same purposes that Health Checking is used, correct? For example, docker/Kubernetes can use the response from the HealthCheck endpoints to detect whether the application is ready to receive requests or if its in some sort of failure state for example. On the other hand, OTEL instrumentation provides metrics from time to time based on each instrumentation and batching settings etc, and doesn't have some "real-time" way of checking the health of the metered components.

Does OTEL itself define/discuss anything health-related? Would it ever make sense to _combine_ these concepts into one, and have health be part of OTEL and be able to be exported to monitoring tools in a similar way?

How are you guys currently handling this? Do you have both things in place, or do you only use OTEL instrumentation?

Thoughts?

**Comment by salvadoriume** [comment]
@julealgon do you have after 2 years any new findings and thoughts? I am now in the same place. I have some checks that I want to attach under kubernetes probes but some of them I want to be "status" checks and I want to transfer as a metric to grafana cloud. The question if it is good approach to use same library for checks to attach them under kubernetes probes and to send them to grafana cloud as metric?

**Reply by julealgon** [reply]
@salvadoriume  I'll admit things have not changed _that much_ from my perspective. We still haven't invested _heavily_ in health checking, but we will very soon.

My own conclusions are:
1. I still believe the concepts of "exporting" (from OTEL) and "publishing" (from health checks) are highly redundant semantically and that there is an opportunity here to unify the systems eventually (up to Microsoft of course, not consumer code).

2. I've been focusing on using health checks for "detecting application-specific problems", and telemetry for "sending constant information about the system". Even as I write this, there is some obvious overlap in the 2: one application might consider "low memory" as a health check indicator, even though telemetry already reports memory constantly. I've avoided _these_ types of checks as health checks and have focused more on things like "can the app connect to the db at all", "did our custom SAML integration load properly or did it fail?", "is our PDF library license properly applied or are we missing settings?", "can the app communicate with Azure AppConfiguration, or did auth or connection fail?". Those vary a lot by application, and they don't have "telemetry equivalents". We are not currently using health publishing capabilities at all: we just have Datadog ping the health endpoints from time to time and warn on issues

One thing you can do if you want to reuse the same information across telemetry and health checking, is you can actually create a meter listener in-memory for some metric that is exported via OTEL, and have the application itself also see that metric and evaluate it. Then you could have your health check endpoint report on the latest "situation" on that metric while not reinventing the entire mechanism to obtain the value in the first place.

Like I mentioned, we are actually avoiding that type of health check for now, but it is certainly possible to model something like that. You could have a hosted service that listens to a metric and performs some aggregation or persistence into a mutable class that is injected as a singleton in both the hosted service and the health check step, allowing the health check to use the value for the decision on whether that is `Healthy`, `Degraded`, etc. This works well because the hosted service is the only writer of that data (and it is a single instance) while the health check only reads it (so multiple readers are fine, in case the health endpoint is somehow hammered).

You can additionally append arbitrary information to the health check step and expose that in the endpoint. On modern projects, we have mostly stopped using the "built-in" health check mapping methods in favor of creating a minimal API endpoint that leverages the `HealthCheckService` class to run the check and then return the report as standard Json. We can then have that custom information from multiple steps be propagated to the API and used by the observability tool to output useful details when the state is unhealthy.

Another reason we opted for "manual" minimal API endpoints for health is that they participate in OpenAPI generation. We need that to be able to update Azure API Management's API surface so that both our deployment process and Datadog can later access that endpoint. The latter does it constantly and we alert on deviations from `Healthy`, while the former checks the health endpoint during blue/green deployment and fails the deployment procedure if the API keeps responding `Unhealthy` for too long.

This is how we are using both mechanisms at the moment. Hope it helps.
